File: test/java_remove_symbols_test.lua
local lib = require("java_symbols_lib")

local function eq(lt, rt)
    for i = 1, #lt do
        if lt[i] ~= rt[i] then
            return false
        end
    end
    
    return true
end

tests = {
function()
    print("test_ignore_whitespace_comments...")
    
    assert(lib.ignore_whitespace_comments([[a]], 1) == 1)
    assert(lib.ignore_whitespace_comments([[ab]], 2) == 2)
    assert(lib.ignore_whitespace_comments([[//]], 1) == 3)
    assert(lib.ignore_whitespace_comments([[/**/]], 1) == 5)
    assert(lib.ignore_whitespace_comments([[/* a */]], 1) == 8)
    assert(lib.ignore_whitespace_comments([[/**/ a]], 1) == 6)
    assert(lib.ignore_whitespace_comments([[//a
]], 1) == 5)
end,

function()
    print("test_next_symbol...")
    
    assert(eq({lib.next_symbol([[]])}, {[[]], 1}))
    assert(eq({lib.next_symbol([[ ]])}, {[[]], 2}))
    assert(eq({lib.next_symbol([[foo]])}, {[[foo]], 4}))
    assert(eq({lib.next_symbol([[ foo ]])}, {[[foo]], 5}))
    assert(eq({lib.next_symbol([[//

foo]])}, {[[foo]], 8}))
    assert(eq({lib.next_symbol("/* */ foo ")}, {[[foo]], 10}))
end,

function()
    print("test_find_token...")
    
    assert(lib.find_token([[@]], "@") == 1)
    assert(lib.find_token([[ @]], "@") == 2)
    assert(lib.find_token([[(@)]], "@") == 2)
    assert(lib.find_token([[//
@]], "@") == 4)
    assert(lib.find_token([[/*
*/
@]], "@") == 7)
    
    assert(lib.find_token([[' '@]], "@") == 4)
    assert(lib.find_token([['\''@]], "@") == 5)
    assert(lib.find_token([['\uFFFE'@]], "@") == 9)
    assert(lib.find_token([["//"@]], "@") == 5)
    assert(lib.find_token([["/*"@]], "@") == 5)
    assert(lib.find_token([[())]], ")") == 3)
    assert(lib.find_token([[()]], ")", 2) == 2)
    assert(lib.find_token([[(()))]], ")") == 5)
    assert(lib.find_token([['"'@]], "@") == 4)
    
    assert(lib.find_token([[// @]], "@") == 5)
    assert(lib.find_token([[// @
]], "@") == 6)
    
    assert(lib.find_token([[/*@*/]], "@") == 6)
    assert(lib.find_token([[/* @ */]], "@") == 8)
    assert(lib.find_token([[/*
@ */]], "@") == 8)
    assert(lib.find_token([[// /*
@]], "@") == 7)
    assert(lib.find_token([[/**//*@ */]], "@") == 11)
    assert(lib.find_token([[/**///@]], "@") == 8)
    
    assert(lib.find_token([['@']], "@") == 4)
    assert(lib.find_token([['\uFFFE']], "\\u") == 9)
    assert(lib.find_token([['\'']], "\\'") == 5)
    assert(lib.find_token([["@"]], "@") == 4)
    assert(lib.find_token([["""@"]], "@") == 6)
    assert(lib.find_token([["" "@"]], "@") == 7)
    assert(lib.find_token([["\\" "@"]], "@") == 9)
    assert(lib.find_token([["\\\"" "@"]], "@") == 11)
    
    assert(lib.find_token([[()]], ")") == 3)
    assert(lib.find_token([[(())]], ")") == 5)
    
    assert(lib.find_token([[noimport]], "import", 1, true) == 9)
    assert(lib.find_token([[_import]], "import", 1, true) == 8)
    assert(lib.find_token([[/import]], "import", 1, true) == 2)
    assert(lib.find_token([[+import]], "import", 1, true) == 2)
    
    assert(lib.find_token([[importnot]], "import", 1, true) == 10)
    assert(lib.find_token([[import_]], "import", 1, true) == 8)
    assert(lib.find_token([[import/]], "import", 1, true) == 1)
    assert(lib.find_token([[import+]], "import", 1, true) == 1)
end,

function()
    print("test_next_annotation...")
    
    assert(eq({lib.next_annotation([[@A]])}, {1, 3, [[A]]}))
    assert(eq({lib.next_annotation([[@A
]])}, {1, 3, [[A]]}))
    assert(eq({lib.next_annotation([[@A()]])}, {1, 5, [[A]]}))
    
    assert(eq({lib.next_annotation([[@A class B {}]])}, {1, 3, [[A]]}))
    assert(eq({lib.next_annotation([[@A(a = ')')]])}, {1, 12, [[A]]}))
    assert(eq({lib.next_annotation([[@A(a = ')') class B {}]])}, {1, 12, [[A]]}))
    assert(eq({lib.next_annotation([[@A(a = ")")]])}, {1, 12, [[A]]}))
    assert(eq({lib.next_annotation([[@A(a = ")))" /*)))*/) class B {}]])}, {1, 22, [[A]]}))
    assert(eq({lib.next_annotation([[@A(/* ) */)]])}, {1, 12, [[A]]}))
    assert(eq({lib.next_annotation([[method(@A Object o)]])}, {8, 10, [[A]]}))
    
    assert(eq({lib.next_annotation([[@A(
value = ")" /* ) */
// )
)
]])}, {1, 31, [[A]]}))
    
    assert(eq({lib.next_annotation([[ // @A
/* @B */
value = "@C";
@D]])}, {31, 33, [[D]]}))
    
    assert(eq({lib.next_annotation([[@a.b.C]])}, {1, 7, [[a.b.C]]}))
    assert(eq({lib.next_annotation([[@a/**/.B]])}, {1, 9, [[a.B]]}))
    
    assert(eq({lib.next_annotation([[@A(value = /* ) */ ")")//)]])}, {1, 24, [[A]]}))
end,

function()
    print("test_remove_imports_simple...")
    
    local original_content = [[
import java.lang.Runnable;
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;]]
    
    assert(lib.remove_imports(original_content, {"Runnable"}) == [[
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;]])
    
    assert(lib.remove_imports(original_content, {"[*]"}) == [[
import java.lang.Runnable;
import java.util.List;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;]])
    
    assert(lib.remove_imports(original_content, {"java[.]util"}) == [[
import java.lang.Runnable;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;]])
    
    assert(lib.remove_imports(original_content, {"util"}) == [[
import java.lang.Runnable;
import static java.lang.String.valueOf;
]])
    
    assert(lib.remove_imports(original_content, {"java"}) == [[
import com.google.common.util.concurrent.Service;]])
    
    assert(lib.remove_imports(original_content, {"static"}) == original_content)
end,

function()
    print("test_remove_imports_whitespace_comments...")
    
    assert(lib.remove_imports([[import A ;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import A ; ]], {"A"}) == [[ ]])
    assert(lib.remove_imports([[import/**/A;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import/**/A/**/;/**/]], {"A"}) == [[/**/]])
    assert(lib.remove_imports([[import//
A;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import A./*B;*/C;]], {"A[.]C"}) == [[]])
    
    assert(lib.remove_imports([[import static A;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import static a . b /**/ . A;]], {"A"}) == [[]])
    
    assert(lib.remove_imports([[import xstatic .A;]], {"static"}) == [[]])
    assert(lib.remove_imports([[import staticx.A;]], {"static"}) == [[]])
    assert(lib.remove_imports([[import static/**/A;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import/**/static/**/A;]], {"A"}) == [[]])
    assert(lib.remove_imports([[import/* A */B;]], {"A"}) == [[import/* A */B;]])
end,

function()
    print("test_remove_annotations_simple...")
    
    assert(lib.remove_annotations([[new @Nullable Object[initialCapacity];]], {"Nullable"}) == [[new Object[initialCapacity];]])
    assert(lib.remove_annotations([[@A(value = /* ) */ ")")//)]], {"A"}) == [[//)]])
    
    assert(lib.remove_annotations([[
@A
class C {}]], {"A"}) == [[class C {}]])
    
    assert(lib.remove_annotations([[
    @A
    class C {}]], {"A"}) == [[    class C {}]])
    
    local original_content = [[
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.Test
@org.junit.jupiter.api.Test]]
    
    assert(lib.remove_annotations(original_content, {"SuppressWarnings"}) == [[
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.Test
@org.junit.jupiter.api.Test]])
    
    assert(lib.remove_annotations(original_content, {"Suppress"}) == [[
@org.junit.Test
@org.junit.jupiter.api.Test]])
    
    assert(lib.remove_annotations(original_content, {"org[.]junit[.]Test"}) == [[
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.jupiter.api.Test]])
    
    assert(lib.remove_annotations(original_content, {"Test"}) == [[
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
]])
    
    assert(lib.remove_annotations(original_content, {"@SuppressWarnings"}) == original_content)
    assert(lib.remove_annotations(original_content, {"EI_EXPOSE_REP"}) == original_content)
end,

function()
    print("test_remove_annotations_whitespace_comments...")
    
    assert(lib.remove_annotations([[@a/*A*/.B]], {"A"}) == [[@a/*A*/.B]])
    assert(lib.remove_annotations([[@a/*A*/.B]], {"B"}) == [[]])
    assert(lib.remove_annotations([[@ A]], {"A"}) == [[]])
    assert(lib.remove_annotations([[@//
A]], {"A"}) == [[]])
    assert(lib.remove_annotations([[@A/*(B)*/]], {"B"}) == [[@A/*(B)*/]])
end,

function()
    print("test_main...")
    
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class1.java", "-n", "D"}) == [[
import a.b.D.A;
import static a.b.C.D;

@D
class Class1 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class1.java", "-a", "-n", "D"}) == [[
import a.b.D.A;
import static a.b.C.D;

class Class1 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-n", "Runnable"}) == [[
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-n", "*"}) == [[
import java.lang.Runnable;
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-n", "valueOf"}) == [[
import java.lang.Runnable;
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-p", "java"}) == [[
import com.google.common.util.concurrent.Service;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-p", "util", "-n", "Runnable"}) == [[
import static java.lang.String.valueOf;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class2.java", "-p", "static"}) == [[
import java.lang.Runnable;
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;

class Class2 {
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class3.java", "-a", "-p", "Test"}) == [[
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
class Class3 {
    public void test1() {
    }
    
    public void test2() {
    }
    
    public void test3() {
    }
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class3.java", "-a", "-n", "Test"}) == [[
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
class Class3 {
    public void test1() {
    }
    
    public void test2() {
    }
    
    public void test3() {
    }
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class3.java", "-a", "-p", "api[.]Test"}) == [[
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
class Class3 {
    public void test1() {
    }
    
    @org.junit.Test
    public void test2() {
    }
    
    public void test3() {
    }
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class3.java", "-a", "-p", "Suppress"}) == [[
import org.junit.jupiter.api.Test;

class Class3 {
    @Test
    public void test1() {
    }
    
    @org.junit.Test
    public void test2() {
    }
    
    @org.junit.jupiter.api.Test
    public void test3() {
    }
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class3.java", "-a", "-p", "EI_EXPOSE_REP"}) == [[
import org.junit.jupiter.api.Test;
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
class Class3 {
    @Test
    public void test1() {
    }
    
    @org.junit.Test
    public void test2() {
    }
    
    @org.junit.jupiter.api.Test
    public void test3() {
    }
}
]])
    assert(lib.main({"--dry-run", "./test/data/java_remove_symbols/Class4.java", "-a", "-n", "Nullable"}) == [[
 /*;*/ // ;

class Class4 {
    <T extends Object> T method(Object o) {
        new Object();
    }
}
]])
end,
}

for _, test in ipairs(tests) do
    test()
end

print("PASS")

File: test/java_remove_symbols_test.cpp
#include <cassert>

#include <java-utils/java_symbols.hpp>

using namespace fedora::java::javapackages::java_utils::java_symbols;

int main(int argc, const char** argv)
{
    using next_annotation_t = std::tuple<std::string_view, std::string>;
    
    assert((ignore_whitespace_comments("a", 0) == 0));
    assert((ignore_whitespace_comments("ab", 1) == 1));
    assert((ignore_whitespace_comments("//", 0) == 2));
    assert((ignore_whitespace_comments("/**/", 0) == 4));
    assert((ignore_whitespace_comments("/* a */", 0) == 7));
    assert((ignore_whitespace_comments("/**/ a", 0) == 5));
    assert((ignore_whitespace_comments("//a\n", 0) == 4));
    
    assert((next_symbol("") == ""));
    assert((next_symbol(" ") == ""));
    assert((next_symbol("(foo") == "("));
    assert((next_symbol("foo") == "foo"));
    assert((next_symbol(" foo ") == "foo"));
    assert((next_symbol("//\n\nfoo") == "foo"));
    assert((next_symbol("/* */ foo ") == "foo"));
    
    assert((find_token("@", "@") == 0));
    assert((find_token(" @", "@") == 1));
    assert((find_token("(@)", "@") == 1));
    assert((find_token("//\n@", "@") == 3));
    assert((find_token("/*\n*/\n@", "@") == 6));
    
    assert((find_token("' '@", "@") == 3));
    assert((find_token("'\''@", "@") == 4));
    assert((find_token("'\\uFFFE'@", "@") == 8));
    assert((find_token("\"//\"@", "@") == 4));
    assert((find_token("\"/*\"@", "@") == 4));
    
    assert((find_token("())", ")") == 2));
    assert((find_token("()", ")", 1) == 1));
    assert((find_token("(()))", ")") == 4));
    assert((find_token("'\"'@", "@") == 3));
    
    assert((find_token("// @", "@") == 4));
    assert((find_token("// @\n", "@") == 5));
    
    assert((find_token("/*@*/", "@") == 5));
    assert((find_token("/* @ */", "@") == 7));
    assert((find_token("/*\n@ */", "@") == 7));
    assert((find_token("// /*@", "@") == 6));
    assert((find_token("/**//*@ */", "@") == 10));
    assert((find_token("/**///@", "@") == 7));
    
    assert((find_token("'@'", "@") == 3));
    assert((find_token(R"('\uFFFE')", R"(\u)") == 8));
    assert((find_token(R"('\'')", R"(\')") == 4));
    assert((find_token(R"("@")", "@") == 3));
    assert((find_token(R"("""@")", "@") == 5));
    assert((find_token(R"("" "@")", "@") == 6));
    assert((find_token(R"("\\" "@")", "@") == 8));
    assert((find_token(R"("\\\"" "@")", "@") == 10));
    
    assert((find_token("()", ")") == 2));
    assert((find_token("(())", ")") == 4));
    
    assert((find_token("noimport", "import", 0, true) == 8));
    assert((find_token("_import", "import", 0, true) == 7));
    assert((find_token("/import", "import", 0, true) == 1));
    assert((find_token("+import", "import", 0, true) == 1));
    
    assert((find_token("importnot", "import", 0, true) == 9));
    assert((find_token("import_", "import", 0, true) == 7));
    assert((find_token("import/", "import", 0, true) == 0));
    assert((find_token("import+", "import", 0, true) == 0));
    
    assert((next_annotation("@A") == next_annotation_t("@A", "A")));
    assert((next_annotation("@A\n") == next_annotation_t("@A", "A")));
    assert((next_annotation("@A()") == next_annotation_t("@A()", "A")));
    
    assert((next_annotation("@A class B {}") == next_annotation_t("@A", "A")));
    
    assert((next_annotation("@A(a = ')')") == next_annotation_t("@A(a = ')')", "A")));
    assert((next_annotation("@A(a = ')') class B {}") == next_annotation_t("@A(a = ')')", "A")));
    
    assert((next_annotation("@A(a = \")\")") == next_annotation_t("@A(a = \")\")", "A")));
    assert((next_annotation("@A(a = \")))\" /*)))*/) class B {}") == next_annotation_t("@A(a = \")))\" /*)))*/)", "A")));
    assert((next_annotation("@A(/* ) */)") == next_annotation_t("@A(/* ) */)", "A")));
    assert((next_annotation("method(@A Object o)") == next_annotation_t("@A", "A")));
    
    assert((next_annotation("@A(\nvalue = \")\" /* ) */\n// )\n)\n") == next_annotation_t("@A(\nvalue = \")\" /* ) */\n// )\n)", "A")));
    
    assert((next_annotation(" // @A\n/* @B */\nvalue = \"@C\";\n@D") == next_annotation_t("@D", "D")));
    
    assert((next_annotation("@a.b.C") == next_annotation_t("@a.b.C", "a.b.C")));
    assert((next_annotation("@a/**/.B") == next_annotation_t("@a/**/.B", "a.B")));
    
    assert((next_annotation("@A(value = /* ) */ \")\")//)") == next_annotation_t("@A(value = /* ) */ \")\")", "A")));
    
    {
        constexpr std::string_view original_content = R"(
import java.lang.Runnable;
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;)";
        
        std::vector<std::regex> args;
        
        args.emplace_back("Runnable");
        assert((std::get<0>(remove_imports(original_content, args, {})) == R"(
import java.util.List;
import static java.util.*;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;)"));
        args.clear();
        
        args.emplace_back("[*]");
        
        assert((std::get<0>(remove_imports(original_content, args, {})) == R"(
import java.lang.Runnable;
import java.util.List;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;)"));
        args.clear();
        
        args.emplace_back("java[.]util");
        assert((std::get<0>(remove_imports(original_content, args, {})) == R"(
import java.lang.Runnable;
import static java.lang.String.valueOf;
import com.google.common.util.concurrent.Service;)"));
        args.clear();
        
        args.emplace_back("util");
        assert((std::get<0>(remove_imports(original_content, args, {})) == R"(
import java.lang.Runnable;
import static java.lang.String.valueOf;
)"));
        args.clear();
        
        args.emplace_back("java");
        assert((std::get<0>(remove_imports(original_content, args, {})) == R"(
import com.google.common.util.concurrent.Service;)"));
        args.clear();
        
        args.emplace_back("static");
        assert((std::get<0>(remove_imports(original_content, args, {})) == original_content));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import A ;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import A ; ", args, {})) == " "));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import/**/A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import/**/A/**/;/**/", args, {})) == "/**/"));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import//\nA;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A[.]C");
        assert((std::get<0>(remove_imports("import A./*B;*/C;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import static A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import static a . b /**/ . A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("static");
        assert((std::get<0>(remove_imports("import xstatic .A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("static");
        assert((std::get<0>(remove_imports("import staticx.A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import static/**/A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import/**/static/**/A;", args, {})) == ""));
        args.clear();
        
        args.emplace_back("A");
        assert((std::get<0>(remove_imports("import/* A */B;", args, {})) == "import/* A */B;"));
        args.clear();
    }
    
    {
        std::vector<std::regex> patterns;
        
        patterns.emplace_back("Nullable");
        assert((remove_annotations("new @Nullable Object[initialCapacity];", patterns, {}, {}) == "new Object[initialCapacity];"));
        patterns.clear();
        
        patterns.emplace_back("A");
        assert((remove_annotations("@A(value = /* ) */ \")\")//)", patterns, {}, {}) == "//)"));
        patterns.clear();
        
        patterns.emplace_back("A");
        assert((remove_annotations(R"(
@A
class C {})", patterns, {}, {}) == "\nclass C {}"));
        patterns.clear();
    
        patterns.emplace_back("A");
        assert((remove_annotations(R"(
    @A
    class C {})", patterns, {}, {}) == "\n    class C {}"));
        patterns.clear();
        
        constexpr std::string_view original_content = R"(
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.Test
@org.junit.jupiter.api.Test)";
        
        patterns.emplace_back("SuppressWarnings");
        assert((remove_annotations(original_content, patterns, {}, {}) == R"(
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.Test
@org.junit.jupiter.api.Test)"));
        patterns.clear();
        
        patterns.emplace_back("Suppress");
        assert((remove_annotations(original_content, patterns, {}, {}) == R"(
@org.junit.Test
@org.junit.jupiter.api.Test)"));
        patterns.clear();
        
        patterns.emplace_back("org[.]junit[.]Test");
        assert((remove_annotations(original_content, patterns, {}, {}) == R"(
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
@org.junit.jupiter.api.Test)"));
        patterns.clear();
        
        patterns.emplace_back("Test");
        assert((remove_annotations(original_content, patterns, {}, {}) == R"(
@SuppressWarnings
@SuppressFBWarnings(value = {"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
)"));
        patterns.clear();
        
        patterns.emplace_back("@SuppressWarnings");
        assert((remove_annotations(original_content, patterns, {}, {}) == original_content));
        patterns.clear();
        
        patterns.emplace_back("EI_EXPOSE_REP");
        assert((remove_annotations(original_content, patterns, {}, {}) == original_content));
        patterns.clear();
        
        patterns.emplace_back("A");
        assert((remove_annotations("@a/*A*/.B", patterns, {}, {}) == "@a/*A*/.B"));
        patterns.clear();
        
        patterns.emplace_back("B");
        assert((remove_annotations("@a/*A*/.B", patterns, {}, {}) == ""));
        patterns.clear();
        
        patterns.emplace_back("A");
        assert((remove_annotations("@ A", patterns, {}, {}) == ""));
        patterns.clear();
        
        patterns.emplace_back("A");
        assert((remove_annotations("@//\nA", patterns, {}, {}) == ""));
        patterns.clear();
        
        patterns.emplace_back("B");
        assert((remove_annotations("@A/*(B)*/", patterns, {}, {}) == "@A/*(B)*/"));
        patterns.clear();
    }
}
